### PR TITLE
Add ClassicalDataStore class to keep track of qubits measured

### LINKED ISSUE #3232: Classical Control: Feedforward
**Problem:** 

Cirq currently only supports a single unrolled circuit.  Measurements results cannot be fed forward to influence the future execution of the circuit.  This leads to, for example, Cirq not being able to easily describe a teleportation experiment.  There are also use cases for simple feed forward in variational ansatz and simple quantum error correcting protocols.

**Depends on feature:** Classical data (#3231)

**Rough requirements:**

- Define how classical data can influence future execution 
- how is control specified
- what is changed by the result of the control
- support of multiple cases
- Fit in with the specification of data in the Classical Data feature.
- Support serialization of the specification of control and of the action resulting from the control.
-

### PR DESCRIPTION
**Pull request by daxfohl**
Adds a `ClassicalDataStore` class so we can keep track of which qubits are associated to which measurements.

Closes #3232. Initially this was created as part 14 (of 14) of https://tinyurl.com/cirq-feedforward to enable qudits in classical conditions, by storing and using dimensions of the measured qubits when calculating the integer value of each measurement when resolving sympy expressions. However it may have broader applicability.

This approach also sets us up to more easily add different types of measurements (#3233, #4274). It will also ease the path to #3002 and #4449., as we can eventually pass this into `Result` rather than the raw `log_of_measurement_results` dictionary. (The return type of `_run` will have to be changed to `Sequence[C;assicalDataStoreReader]`.

Related: #887, #3231 (open question @95-martin-orion whether this closes those or not)

This PR contains a `ClassicalDataStoreReader` and `ClassicalDataStoreBase` parent "interface" for the `ClassicalDataStore` class as well. This will allow us to swap in different representations that may have different performance characteristics. See #3808 for an example use case. This could be done by adding an optional `ClassicalDataStore` factory method argument to the `SimulatorBase` initializer, or separately to sampler classes.

(Note this is an alternative to #4778 for supporting qudits in sympy classical control expressions, as discussed here: https://github.com/quantumlib/Cirq/pull/4778/files#r774816995. The other PR was simpler and less invasive, but a bit hacky. I felt even though bigger, this seemed like the better approach and especially fits better with our future direction, and closed the other one).

**Breaking Changes**:
1. The abstract method `SimulatorBase._create_partial_act_on_args` argument `log_of_measurement_results: Dict` has been changed to `classical_data: ClassicalData`. Any third-party simulators that inherit `SimulatorBase` will need to update their implementation accordingly.
2. The abstract base class `ActOnArgs.__init__` argument `log_of_measurement_results: Dict` is now copied before use. For users that depend on the pass-by-reference semantics (this should be rare), they can use the new `classical_data: ClassicalData` argument instead, which is pass-by-reference.

### CODE REVIEW

**Comment by daxfohl** [comment]
I went ahead and created a ClassicalDataReader and ClassicalDataBase parent "interface" for the ClassicalData class as well. This will allow us to swap in different representations that may have different performance characteristics. See #3808 and #4274.

File: cirq-core/cirq/value/classical_data.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        self,
+        *,
+        _measurements: Dict['cirq.MeasurementKey', Tuple[int, ...]] = None,
+        _measured_qubits: Dict['cirq.MeasurementKey', Tuple['cirq.Qid', ...]] = None,
+        _channel_measurements: Dict['cirq.MeasurementKey', int] = None,
+        _measurement_types: Dict['cirq.MeasurementKey', 'cirq.MeasurementType'] = None,
+    ):
+        """Initializes a `ClassicalDataDictionaryStore` object."""
+        _measurement_types_was_none = _measurement_types is None
+        if _measurement_types is None:
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/classical_data.py:117]: This section would be cleaner using an else block:
```
added_types = {}
if _measurement_types is None:
    _measurement_types = {}
else:
    # populate added_types
_measurement_types.update(added_types)
```

File: cirq-core/cirq/value/classical_data.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+
+@value_equality_attr.value_equality(unhashable=True)
+class ClassicalDataDictionaryStore(ClassicalDataStore):
+    """Classical data representing measurements and metadata."""
+
+    _measurements: Dict['cirq.MeasurementKey', Tuple[int, ...]]
+    _measured_qubits: Dict['cirq.MeasurementKey', Tuple['cirq.Qid', ...]]
+    _channel_measurements: Dict['cirq.MeasurementKey', int]
+    _measurement_types: Dict['cirq.MeasurementKey', 'cirq.MeasurementType']
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/classical_data.py:105]: These look like they can be removed (maybe leftover from dataclass-formatting?)

File: cirq-core/cirq/value/classical_data.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+import enum
+from typing import Dict, Mapping, Sequence, Tuple, TYPE_CHECKING, TypeVar
+
+from cirq.value import digits, value_equality_attr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class MeasurementType(enum.IntEnum):
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/classical_data.py:25]: Do you know how this interacts with [previous issues with enum serialization](https://github.com/quantumlib/Cirq/issues/3853) that we've seen?

File: cirq-core/cirq/value/classical_data.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    def copy(self):
+        """Creates a copy of the object."""
+        return ClassicalDataDictionaryStore(
+            _measurements=self._measurements.copy(),
+            _measured_qubits=self._measured_qubits.copy(),
+            _channel_measurements=self._channel_measurements.copy(),
+            _measurement_types=self._measurement_types.copy(),
+        )
+
+    def _json_dict_(self):
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/classical_data.py:245]: For consistent serialization behavior, these lists should be sorted.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Some initial comments on the ClassicalData types.

File: cirq-core/cirq/value/classical_data.py (hunk 4 above)
Comment (daxfohl) [reply cirq-core/cirq/value/classical_data.py:245]: This was intentional. For python 3.6 up, dictionaries preserve insertion order, so this should be consistent. And I think the order of insertion would be useful to preserve on serialization too, especially since we expose a `keys` method.

File: cirq-core/cirq/value/classical_data.py (hunk 3 above)
Comment (daxfohl) [reply cirq-core/cirq/value/classical_data.py:25]: IntEnum serializes to an integer (and deserializes too). It took a while to figure out a way to get it to all work, but this seems to do everything we need it to. I may have missed something, but I *think* this is a valid approach. LMK if you see anything I missed.

File: cirq-core/cirq/value/classical_data.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+                measured or if the measurement key was already used.
+        """
+
+    @abc.abstractmethod
+    def record_channel_measurement(self, key: 'cirq.MeasurementKey', measurement: int):
+        """Records a channel measurement.
+
+        Args:
+            key: The measurement key to hold the measurement.
+            measurement: The measurement result.
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/value/classical_data.py:91]: Maybe the Kraus probabilities? I didn't think the qubits were relevant here so didn't include them, figuring it would create confusion that the value here could be larger than the product of qid dimensions. I figured if we determined that we do need something, it's easy to add later?

File: cirq-core/cirq/value/classical_data.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+
+class ClassicalDataStoreReader(abc.ABC):
+    @abc.abstractmethod
+    def keys(self) -> Tuple['cirq.MeasurementKey', ...]:
+        """Gets the measurement keys in the order they were stored."""
+
+    @abc.abstractmethod
+    def get_int(self, key: 'cirq.MeasurementKey') -> int:
+        """Gets the integer corresponding to the measurement.
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/value/classical_data.py:43]: I updated the docstrings. Your second guess was correct about get_digits. Maybe should that be called get_measurements? Is there a better name for get_int?

**Comment by daxfohl** [comment]
@95-martin-orion should we consider https://github.com/quantumlib/Cirq/issues/3708 here?

**Comment by 95-martin-orion** [comment]
> @95-martin-orion should we consider #3708 here?

Probably not - I don't think we came to a conclusive agreement that #3708 should be done in the first place. I've flagged the related #3868 to be revisited before Cirq 1.0 so we make sure to come back to this.

**Comment by daxfohl** [comment]
A quick synopsis of some rationale in some details in the PR:

1. Added `_create_partial_act_on_args_ex` instead of modifying the existing `_create_partial_act_on_args`, for backwards compatibility of any third party that may have already implemented another SimulatorBase subclass.
2. Removed `@abstractmethod` from `_create_partial_act_on_args` and did not add it to `..._ex`, for forward and backward compatibility. (Just now realized an `implements_one_of` attribution could be useful here. Though I plan on deprecating the old method so the attribute may be overkill. Probably `..._ex` should have a better long-term name).
3. ActOnArgs classes all still accept a `log_of_measurement_results` in `__init__`, for backwards compatibility and use it to initialize the `classical_data` but I do plan on deprecating this.
4. ActOnArgs classes all still provide a `log_of_measurement_results` (via its OperationTarget base class) for backwards compatibility, but I plan on deprecating that too. These aren't big deprecations internally, but affect a ton of unit tests.
4. I considered separating this PR into two: one for integrating classical_data, and then adding sympy qudit support separately. However, it turns out that doing the first part implicitly makes the second part work; I'd have to intentionally add a bug.

**Comment by maffoo** [comment]
This looks interesting, but it seems to me that we might want to integrate this more closely with `cirq.Result`. AFAICT, a `ClassicalDataStore` is something like a "builder" for `cirq.Result`, and I think structuring things in that way could be very beneficial. I will note that as part of working on removing the uniqueness constraint on measurement keys (#4274) I'm trying to separate `cirq.Result` into separate interface and implementation classes, as in `ClassicalDataStore` and `ClassicalDataDictionaryStore` here.

**Comment by daxfohl** [comment]
@maffoo The biggest difference is that `ClassicalDataStore` is for a single simulation, whereas `Result` is for a run (multiple repetitions or samplers). Well, and the other difference is that `ClassicalDataStore` is also *used* during simulation, for simulating classically controlled gates, so merging the two concepts together and force simulation logic to be repetition-aware, which seems out-of-place.

The way I was imagining, we'd update the constructor of `Result` to accept a `Sequence[ClassicalDataStoreReader]`, or possibly have a subclass of `Result` that does this.

For dropping the uniqueness constraint on measurements, I was imagining we would add a new function `ClassicalDataStore.record_repeated_measurement(bits, qubits_measured)` that would put those in a separate dictionary, of type `Dict[MeasurementKey, List[Tuple[int, ...]]`. It should also store the qubits_measured and raise a ValueError if the qubits_measured is not the same shape as the previous measurement on the same key. We would also add a new reader function `ClassicalDataStoreReader.get_repeated_measurement` that would return the array of measurements at that key. Then `Result` (or subclass thereof) could use this however it likes.

That said I do worry about the possibility of exploding different reader/writer methods here if we start doing lots of different measurement types. Maybe a single `record_measurement(key: MeasurementKey, value: Any)` is better? My gut says it's better to have specific readers/writers for each type (single measurements, repeated measurements, channel measurements) and that these won't explode beyond a couple, but it's hard to be sure.

(I was thinking MeasurementGate would have a new property allow_repeated, to signal whether this was a repeated measurement or not, and that ClassicallyControlledGate could have a new `index` property to control based on a repeated measurement offset. But maybe you're planning something completely different).

Anyway it probably makes sense to review what you're planning for #4274 before moving forward with this, to make sure we don't have to redo it all in a month.

**Comment by daxfohl** [comment]
@maffoo Can you help me understand the concern with having a `class RepeatedSimulationResult(Result)` that accepts `Sequence[ClassicalDataStoreReader]` (one for each repetition)?

The issue I'm having in any design that uses a single data store object for all repetitions, is that this forces simulation to be repetition-aware, which ends up being quite invasive. Also it makes the `.simulate()` function odd, in that the classical data returned would be indexed by repetition, though there are no repetitions. (We could write a wrapper around that, that gets the zero index of everything, but that seems upside-down).

One thing that *would* be straightforward, if performance is a concern, is to special-case a Result type for when we're sampling from circuits that have only terminal measurements. That goes through a completely different code path (it just calculates the resulting state vector or density matrix and samples it N times -- circuits also have to be unitary / noiseless to go down this code path in state vector simulators), so we could return a different `Result` subclass in those cases, that may be more efficient. 

However for things that go through the full simulation on each repetition, forcing repetition-awareness into it would be pretty big. That said, if the concern is performance here, I'd expect the repeated execution of the simulations themselves would dominate the CPU time; I'd not expect aggregating the measurement dictionaries to add much time at all comparatively. Very high level pseudocode would be like:

```python
class SimulatorBase:
  ...
  def sample(count: int) -> Result:
    if can_run_as_sampler(circuit):
      v = get_state_vector(circuit)
      dict = v.sample(count) # returns a Dict[MeasurementKey, List of size `count`]
      return PreAggregatedDictResult(dict)
    else:
      results = [simulate(circuit).measurements for _ in range(count)]
      return ResultFromRepeatedSimulations(results) # or we could just aggregate it directly in this function
```


Anyway I'm wanting to understand what the primary concern is before writing up a full doc with this invasive of a change, to make sure I'm capturing all the right aspects.  cc @95-martin-orion @MichaelBroughton

**Comment by daxfohl** [comment]
@95-martin-orion @maffoo This is going stale. My take is that this is the most straightforward design, in that one datastore is one simulation. This is ideal from the classical control perspective, though maybe less so in the sampling perspective.

But this also provides interfaces such that if we want to create alternate implementations that have better performance with repeated simulations, it's possible to add those later without breaking anything. Can we proceed with this or do we need more feedback?

**Comment by 95-martin-orion** [comment]
@daxfohl Thanks for your patience on this. My read on this is that using `Sequence[ClassicalDataStoreReader]` as the source for building `Result`s doesn't conflict with the changes @maffoo is looking at for reusable keys - the fact that `Result` intentionally uses repetitions as the outermost index on its storage simplifies this.

Rather than making simulations repetition-aware (which I agree would be a wide-sweeping change), I support continuing with the original design of this PR. I'll follow up with additional review either later today or Monday.

File: cirq-core/cirq/sim/clifford/clifford_simulator_test.py [hunk 7]
```diff
@@ -546,7 +546,7 @@ def test_valid_apply_measurement():
     state = cirq.CliffordState(qubit_map={q0: 0}, initial_state=1)
     measurements = {}
     _ = state.apply_measurement(cirq.measure(q0), measurements, np.random.RandomState())
-    assert measurements == {'0': [1]}
+    assert measurements == {'0': (1,)}
```
Comment (95-martin-orion) [inline cirq-core/cirq/sim/clifford/clifford_simulator_test.py:549]: While the behavior of `[1]` and `(1,)` are mostly the same, this feels like a red flag to me. Why does the type of `measurements` change, and how does this affect other parts of Cirq?

File: cirq-core/cirq/sim/operation_target.py [hunk 8]
```diff
@@ -86,9 +86,14 @@ def qubits(self) -> Tuple['cirq.Qid', ...]:
         """Gets the qubit order maintained by this target."""
 
     @property
-    @abc.abstractmethod
     def log_of_measurement_results(self) -> Dict[str, Any]:
         """Gets the log of measurement results."""
+        return {str(k): list(self.classical_data.get_digits(k)) for k in self.classical_data.keys()}
```
Comment (95-martin-orion) [inline cirq-core/cirq/sim/operation_target.py:91]: [No change needed] This is OK, but it's pushing the upper limit of "amount of work allowed to be done in a `@property`."

File: cirq-core/cirq/value/classical_data_test.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def test_record_channel_measurement():
+    cd = cirq.ClassicalDataDictionaryStore()
+    cd.record_channel_measurement(mkey_m, 1)
+    assert cd.channel_measurements == {mkey_m: 1}
+    assert cd.keys() == (mkey_m,)
+
+
+def test_record_channel_measurement_errors():
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/classical_data_test.py:70]: Add tests for cross-type errors (i.e. overwriting "normal" measurement with channel measurement and vice-versa)

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
All files reviewed.

File: cirq-core/cirq/contrib/quimb/mps_simulator_test.py [hunk 10]
```diff
@@ -550,10 +550,10 @@ def test_state_act_on_args_initializer():
     s = ccq.mps_simulator.MPSState(
         qubits=(cirq.LineQubit(0),),
         prng=np.random.RandomState(0),
-        log_of_measurement_results={'test': 4},
+        log_of_measurement_results={'test': [4]},
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/contrib/quimb/mps_simulator_test.py:553]: Yes, the test just pushed the value from input straight to output.

File: cirq-core/cirq/sim/act_on_args.py [hunk 11]
```diff
@@ -215,8 +218,8 @@ def _on_transpose_to_qubit_order(self: TSelf, qubits: Sequence['cirq.Qid'], targ
         functionality, if supported."""
 
     @property
-    def log_of_measurement_results(self) -> Dict[str, List[int]]:
-        return self._log_of_measurement_results
+    def classical_data(self) -> 'cirq.ClassicalDataStoreReader':
+        return self._classical_data
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:222]: I removed it from `ActOnArgs` and `ActOnArgsContainer` and moved it to the base class `OperationTarget`.

File: cirq-core/cirq/sim/operation_target.py (hunk 8 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/operation_target.py:91]: This is to support backwards compatibility. I moved this from `ActOnArgs` and `ActOnArgsContainer` to the base class `OperationTarget`.

Comment (daxfohl) [reply cirq-core/cirq/sim/operation_target.py:91]: (Also I think we should deprecate this property next. Currently it's used in `SimulatorBase._run` to create the `cirq.Result`. However it would make more sense if we created a `cirq.Result` constructor/factory/subclass that accepted `ClassicalData` objects directly.)

File: cirq-core/cirq/sim/clifford/clifford_simulator_test.py (hunk 7 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/clifford/clifford_simulator_test.py:549]: Yeah, this bugs me too. In `ClassicalData` I store measurements as tuples, just because immutability. Mostly measurements are accessed via `log_of_measurement_results` so I change those tuples to lists there for backwards compatibility. This is the one other place that exposes measurements and I couldn't initially figure out a way to make this one backwards compatible, and it doesn't seem like a function that would be used by anyone outside of cirq itself. I'll try again to see if there's a way to make this backwards compatible too though. Otherwise we can either revert back to List for measurements, or accept the minor breaking change. (Sounds like immutability isn't a big thing in the Python ecosystem anyway).

File: cirq-core/cirq/value/classical_data.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+            if _channel_measurements:
+                _measurement_types.update(
+                    {k: MeasurementType.CHANNEL for k, v in _channel_measurements.items()}
+                )
+        if _measurements is None:
+            _measurements = {}
+        if _measured_qubits is None:
+            _measured_qubits = {}
+        if _channel_measurements is None:
+            _channel_measurements = {}
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/value/classical_data.py:135]: It changes the behavior. `{}` is False, so it is replaced by a new `{}` if we go with the shorter statement. This is probably fine, but I think better to be consistent in that if user passes in a dict then that dict is what is used.

File: cirq-core/cirq/sim/act_on_args.py [hunk 13]
```diff
@@ -103,9 +106,9 @@ def measure(self, qubits: Sequence['cirq.Qid'], key: str, invert_mask: Sequence[
             return
         bits = self._perform_measurement(qubits)
         corrected = [bit ^ (bit < 2 and mask) for bit, mask in zip(bits, invert_mask)]
-        if key in self._log_of_measurement_results:
-            raise ValueError(f"Measurement already logged to key {key!r}")
-        self._log_of_measurement_results[key] = corrected
+        self._classical_data.record_measurement(
+            value.MeasurementKey.parse_serialized(key), corrected, qubits
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py:110]: The key is serialized. The function signature is `def measure(self, qubits: Sequence['cirq.Qid'], key: str, invert_mask: Sequence[bool])`. Maybe github collapsed it so it looked like it was coming from somewhere else.

[the remark this answers is not in the document]
Comment (95-martin-orion) [reply cirq-core/cirq/sim/act_on_args.py:110]: I think I misunderstood the intent of `parse_serialized` - it doesn't refer to JSON serialization. This is OK.

File: cirq-core/cirq/sim/act_on_args.py [hunk 14]
```diff
@@ -70,11 +73,11 @@ def __init__(
             prng = cast(np.random.RandomState, np.random)
         if qubits is None:
             qubits = ()
-        if log_of_measurement_results is None:
-            log_of_measurement_results = {}
         self._set_qubits(qubits)
         self.prng = prng
-        self._log_of_measurement_results = log_of_measurement_results
+        # pylint: disable=line-too-long
+        self._classical_data = classical_data or value.ClassicalDataDictionaryStore(_measurements=log_of_measurement_results)  # type: ignore
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py]: No, this is a good catch. I thought it was complaining about the `Optional`. I fixed this by making a copy of `log_of_measurement_results` with the required format. **HOWEVER** making a copy here strictly changes the behavior. Anyone dependent on `log_of_measurement_results` being passed by reference is now broken.

Note specifically that `stabilizer_ch_form_test.test_run` needed to be changed because of this. So if we think users might be doing the thing in that test, then we should worry about this. If not, then we shouldn't. (Since that test directly creates StabilizerStateChForm and ActOnStabilizerChFormArgs rather than just running CliffordSimulator, I'd assume this isn't a real-world use-case).

I could maybe mitigate this by temporarily adding a `old_log_of_measurement_results` argument/field to `ClassicalDataDictionaryStore` and duplicate the data there. Not sure it's worth it but I believe it would work.

File: cirq-core/cirq/value/measurement_key.py [hunk 15]
```diff
@@ -77,6 +77,13 @@ def __hash__(self):
             object.__setattr__(self, '_hash', hash(str(self)))
         return self._hash
 
+    def __lt__(self, other):
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/value/measurement_key.py:80]: Now it does.

File: cirq-core/cirq/sim/act_on_args.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/act_on_args.py]: The pass-by-reference is required for separable state simulation. Each substate needs to refer to the same measurement data container. This PR changes that container from being the raw dictionary to being the ClassicalData object, and the simulators have been changed to share that object when states are separable (in create_partial_act_on_args).

The only real-world case in which this would break end users is if all of the following hold:
1. They have a custom simulator that extends `SimulatorBase`.
2. That simulator is capable of representing separable states via `ActOnArgsContainer` and `ActOnArgs.kron/factor/transpose`.
3. The circuits they're running on it perform measurements on one substate, and then do a classical control based on that measurement on a separate substate. I *think* this would end up breaking, but not even sure of that.

The fix for them is to update `create_partial_act_on_args` for their simulator as we did in this PR. (Which we decided was an allowed breaking change anyway).

TLDR: yeah I think we can mark it as breaking and move on.

**Comment by daxfohl** [comment]
@95-martin-orion I think try to get #4847 merged before this one, as it'll create some conflicts for the new joiner otherwise.

**Comment by 95-martin-orion** [comment]
> @95-martin-orion I think try to get #4847 merged before this one, as it'll create some conflicts for the new joiner otherwise.

#4847 is now merged.

**Review by 95-martin-orion — APPROVED** [review]
All of my comments have been addressed.

To help whoever handles the next release, could you edit the first comment in the PR to also include notes on what is broken by this change? Thanks!

**Comment by daxfohl** [comment]
@95-martin-orion added, lmk if you think it needs more details.

**Comment by 95-martin-orion** [comment]
Looks good to me! To put my 2¢ in on why these breaking changes are OK:

1. `SimulatorBase` is an aggregation of behavior shared by Cirq-internal simulators. External simulators are not expected to inherit from it, and are instead directed to inherit from the `SimulatesFoo` interfaces.
2. ActOnArgs is an internal type for handling simulator state. I don't expect references to this type outside of Cirq, much less reliance on specific behavior of one of its arguments.

...so on the distant chance that you're reading this and were broken by this change, feel free to CC me on the issue.
